### PR TITLE
Add MuonHyperball Optimizer

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
## MuonHyperball Optimizer

**MuonHyperball** extends the Muon optimizer with hyperball-style norm-preserving weight updates. It performs gradient descent on the sphere manifold while preserving the weight norm throughout training.

> ⚠️ **Warning**: This optimizer is experimental and may change in future versions.

### Update Rule

The optimizer follows this update rule:

$$W_{t+1} = R \cdot \text{normalize}(W_t - \text{lr} \cdot R \cdot \text{normalize}(\text{update}))$$

where $R$ is the Frobenius norm of $W_t$ (or a user-specified radius). This keeps the weight matrix at constant scale while updating.

### Key Features

- **Norm Preservation**: Weight matrices maintain their Frobenius norm after each optimizer step
- **Fixed Radius Option**: Optionally specify a fixed radius to rescale all parameters at initialization
- **Inherits from Muon**: All Muon features (Newton-Schulz orthogonalization, spectral scaling, etc.) are available

### Usage

from emerging_optimizers.orthogonalized_optimizers import muon_hyperball

# Use parameter norms as radius (default)
optimizer = muon_hyperball.MuonHyperball(model.parameters(), lr=3e-4)

# Use fixed radius for all parameters
optimizer = muon_hyperball.MuonHyperball(model.parameters(), lr=3e-4, hyperball_radius=1.0)

### CODE REVIEW

File: emerging_optimizers/orthogonalized_optimizers/orthogonalized_optimizer.py [hunk 1]
```diff
@@ -206,5 +207,29 @@ def orthogonalize(self, p: torch.Tensor, grad: torch.Tensor, **kwargs: Any) -> t
         grad = self.scaled_orthogonalize_fn(grad)
         return grad
 
+    def pre_weight_update(self, p: torch.Tensor, update: torch.Tensor) -> None:
+        """Hook called before the final weight update.
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/orthogonalized_optimizer.py]: don't call it hook, it is not. Just say function.
is it inplace? if so, follow the convention to call it inplace, e.g. `pre_weight_update_fn_inplace`, same for the post one

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+__all__ = ["MuonHyperball"]
+
+
+class MuonHyperball(Muon):
+    """Muon optimizer with hyperball-style norm-preserving weight updates.
+
+    This optimizer extends Muon by performing gradient descent on the sphere manifold
+    while preserving the weight norm. The update rule is:
+
+        W_{t+1} = R * normalize(W_t - lr * R * normalize(update))
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py]: missing math syntax? make sure latex equation shows correctly

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    where R is the Frobenius norm of W_t. This keeps the weight matrix at constant
+    scale while updating.
+
+    See :class:`~emerging_optimizers.orthogonalized_optimizers.muon.Muon` for full documentation
+    of the base Muon optimizer.
+    """
+
+    def __init__(self, *args: Any, hyperball_eps: float = 1e-8, **kwargs: Any) -> None:
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py]: force hyperball_eps keyword argument, or just pop it out from kwargs.
having single positional argument between args and kwargs is error prone.

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py (hunk 2 above)
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py]: give warning that it is experimental.

File: emerging_optimizers/orthogonalized_optimizers/orthogonalized_optimizer.py [hunk 4]
```diff
@@ -206,5 +207,29 @@ def orthogonalize(self, p: torch.Tensor, grad: torch.Tensor, **kwargs: Any) -> t
...
+        Subclasses can override this to implement custom behavior before the weight update.
+        For example, to implement hyperball-style updates that preserve weight norms.
+
+        Args:
+            p: The parameter tensor.
+            update: The orthogonalized gradient tensor (will be applied as p -= lr * update).
+        """
+        pass
+
+    def post_weight_update(self, p: torch.Tensor, update: torch.Tensor) -> None:
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/orthogonalized_optimizer.py]: add warning in docstring say both are experimental.

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, override
+
+import torch
+
+from emerging_optimizers.orthogonalized_optimizers.muon import Muon
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py]: Nit: This is OK. But try to follow Google style more, import module not class.

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    See :class:`~emerging_optimizers.orthogonalized_optimizers.muon.Muon` for full documentation
+    of the base Muon optimizer.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialize MuonHyperball optimizer.
+
+        Args:
+            *args: Arguments passed to Muon.
+            hyperball_eps: Epsilon for numerical stability in normalization. Defaults to 1e-8.
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py]: move it to "Keyword argument"

File: emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        self.hyperball_radius = hyperball_radius
+        super().__init__(*args, **kwargs)
+
+        # Validate and optionally rescale parameters based on hyperball_radius.
+        with torch.no_grad():
+            for group in self.param_groups:
+                for p in group["params"]:
+                    p_norm = p.norm()
+                    # Validate that parameter has non-zero norm.
+                    if p_norm.item() == 0:
```
Comment (skyw) [inline emerging_optimizers/orthogonalized_optimizers/muon_hyperball.py:76]: Just an FYI, this triggers a device sync and can have profound impact on performance.

We can leave it in the experimental API for now.

**Review by skyw — APPROVED** [review]
